edgelet/iotedged/src/lib.rs

```diff
@@ -264,49 +280,80 @@ impl Main {
                     let code = start_api(
                         &settings,
                         hyper_client.clone(),
-                        &runtime,
-                        &key_store,
+                        &$runtime,
+                        &$key_store,
                         cfg.clone(),
-                        root_key.clone(),
+                        $root_key.clone(),
                         make_shutdown_signal(),
                         &crypto,
                         &mut tokio_runtime,
                     )?;
                     code == StartApiReturnStatus::Restart
                 } {}
+            }};
+        }
+
+        info!("Provisioning edge device...");
+        match settings.provisioning() {
+            Provisioning::Manual(manual) => {
+                info!("Starting provisioning edge device via manual mode...");
+                let (key_store, provisioning_result, root_key) =
+                    manual_provision(&manual, &mut tokio_runtime)?;
+                start_edgelet!(key_store, provisioning_result, root_key, runtime);
+            }
+            Provisioning::External(external) => {
+                info!("Starting provisioning edge device via external provisioning mode...");
+                let external_provisioning_client =
+                    ExternalProvisioningClient::new(external.endpoint()).context(
+                        ErrorKind::Initialize(InitializeErrorReason::ExternalProvisioningClient),
+                    )?;
+                let external_provisioning = ExternalProvisioning::new(external_provisioning_client);
+
+                let provision_fut = external_provisioning
+                    .provision(MemoryKeyStore::new())
+                    .map_err(|err| {
+                        Error::from(err.context(ErrorKind::Initialize(
+                            InitializeErrorReason::ExternalProvisioningClient,
+                        )))
+                    });
+
+                let provisioning = tokio_runtime.block_on(provision_fut);
+
+                provisioning.and_then(move |prov_result| {
+                    prov_result
+                        .credentials()
+                        .map_or_else(
+                            || {
+                                info!("Credentials are expected to be populated for external provisioning.");
+                                Err(Error::from(ErrorKind::Initialize(
+                                    InitializeErrorReason::ExternalProvisioningClient,
+                                )))
+                            },
+                            |credentials| {
+                                if let AuthType::SymmetricKey(symmetric_key) = credentials.auth_type() {
```

Comment:
Would be better to make this a `match` and change the `else` to an explicit `AuthType::X509(_)` arm, so that it guards against future `AuthType`s being added without updating this code.